src/main/java/io/cdap/plugin/gcp/bigquery/sink/DelegatingMultiSinkRecordWriter.java

```diff
@@ -0,0 +1,114 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.gcp.bigquery.sink;
+
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.schema.Schema;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapreduce.OutputCommitter;
+import org.apache.hadoop.mapreduce.RecordWriter;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Record Writer which delegates writes to other Record Writers based on the record's Table name.
+ * <p>
+ * This Record Writer will initialize record writes and Output Committers as needed.
+ */
+public class DelegatingMultiSinkRecordWriter extends RecordWriter<StructuredRecord, NullWritable> {
+
+  private final TaskAttemptContext initialContext;
+  private final String tableNameField;
+  private final String bucketName;
+  private final String bucketPathUniqueId;
+  private final String projectName;
+  private final String datasetName;
+  private final Map<String, RecordWriter<StructuredRecord, NullWritable>> delegateMap;
+  private final DelegatingMultiSinkOutputCommitter delegatingOutputCommitter;
+
+  public DelegatingMultiSinkRecordWriter(TaskAttemptContext initialContext,
+                                         String tableNameField,
+                                         String bucketName,
+                                         String bucketPathUniqueId,
+                                         String projectName,
+                                         String datasetName,
+                                         DelegatingMultiSinkOutputCommitter delegatingMultiSinkOutputCommitter) {
+    this.initialContext = initialContext;
+    this.tableNameField = tableNameField;
+    this.bucketName = bucketName;
+    this.bucketPathUniqueId = bucketPathUniqueId;
+    this.projectName = projectName;
+    this.datasetName = datasetName;
+    this.delegateMap = new HashMap<>();
+    this.delegatingOutputCommitter = delegatingMultiSinkOutputCommitter;
+  }
+
+  @Override
+  public void write(StructuredRecord key, NullWritable value) throws IOException, InterruptedException {
+    String tableName = key.get(tableNameField);
+
+    RecordWriter<StructuredRecord, NullWritable> delegate;
+
+    if (delegateMap.containsKey(tableName)) {
+      delegate = delegateMap.get(tableName);
+    } else {
+      delegate = getRecordWriterDelegate(tableName, key.getSchema());
+    }
+
+    delegate.write(key, value);
+  }
+
+  @Override
+  public void close(TaskAttemptContext context) throws IOException, InterruptedException {
```

Comment:
@albertshau I realized that, for some reason, the Context instance that is supplied to this method does not include all the configuration options that I expect.

If I commit the job with the original instance that was used to create the committer and record writer, the OutputCommitter has everything it needs to commit and update schema.

I wonder if you have any cues as to why the instance of the TaskAttemptContext supplied here would be missing some fields.

Reply:
hm that's odd, I'm not sure why that would be the case. Worth adding a comment though, in case somebody wonders why it's being done the way it is.

Reply:
I'll add a comment.